File: apps/browser-demo/src/deviceAgent/__tests__/anthropicProvider.test.ts
// Ported from
// apps/android-twa/app/src/test/java/com/agentic/wallet/agent/provider/AnthropicProviderTest.kt
// with one inverted assertion: the browser-native provider MUST send
// anthropic-dangerous-direct-browser-access: 'true' (Kotlin asserts its
// absence; we assert its presence because Anthropic's CORS policy blocks
// browser requests without it).

import { describe, expect, it } from 'vitest';

import { buildPlanMessages, buildReviewMessages } from '../prompts/messageAssembler.js';
import { DEVICE_AGENT_SYSTEM_PROMPTS } from '../prompts/systemPrompts.js';
import type { RuntimeConfig } from '../runtime/config.js';

import { AnthropicProvider } from '../provider/anthropicProvider.js';
import { ProviderHttpError } from '../provider/errorCodes.js';

import { FakeHttpExecutor } from './fakeHttpExecutor.helper.js';

function config(model = 'claude-opus-4-5'): RuntimeConfig {
  return {
    provider: 'anthropic',
    apiFormat: 'anthropic',
    model,
    baseUrl: 'https://api.anthropic.com',
    apiKey: 'sk-ant-test-ABCDEFGHIJKLMNOP',
  };
}

describe('AnthropicProvider.generatePlan', () => {
  it('sends the system + user roles, anthropic-version, x-api-key, and the browser-direct header', async () => {
    const http = new FakeHttpExecutor();
    http.queueResponse(
      200,
      JSON.stringify({
        content: [
          {
            type: 'text',
            text:
              '{"intent":"swap","route":"jupiter","risk":"low","approval":"once","safeguards":["check slippage"]}',
          },
        ],
      }),
    );
    const provider = new AnthropicProvider(config(), http);
    const result = await provider.generatePlan({ userPrompt: 'swap 1 SOL' });

    expect(result.intent).toBe('swap');
    expect(http.calls.length).toBe(1);
    const call = http.calls[0]!;
    expect(call.url).toBe('https://api.anthropic.com/v1/messages');
    expect(call.headers['x-api-key']).toBe('sk-ant-test-ABCDEFGHIJKLMNOP');
    expect(call.headers['anthropic-version']).toBe('2023-06-01');
    // The deliberate browser divergence — Kotlin asserts the absence; we assert presence.
    expect(call.headers['anthropic-dangerous-direct-browser-access']).toBe('true');

    const body = JSON.parse(call.body) as Record<string, unknown>;
    expect(body.model).toBe('claude-opus-4-5');
    expect(body.max_tokens).toBe(1024);
    expect(body.temperature).toBe(0.2);
    expect(body.system).toBe(DEVICE_AGENT_SYSTEM_PROMPTS.PLAN);
    const messages = body.messages as Array<{ role: string; content: string }>;
    expect(messages.length).toBe(1);
    expect(messages[0]!.role).toBe('user');
    expect('tools' in body).toBe(false);
    expect('response_format' in body).toBe(false);
  });

  it('user content matches the Phase 4 message assembler output exactly', async () => {
    const http = new FakeHttpExecutor();
    http.queueResponse(200, JSON.stringify({ content: [{ type: 'text', text: '{"intent":"x"}' }] }));
    const provider = new AnthropicProvider(config(), http);
    const payload = { userPrompt: 'swap 1 SOL', userNotes: 'make it fast' };
    await provider.generatePlan(payload);

    const body = JSON.parse(http.calls[0]!.body) as Record<string, unknown>;
    const messages = body.messages as Array<{ role: string; content: string }>;
    expect(messages[0]!.content).toBe(buildPlanMessages(payload).userContent);
  });

  it('429 → ProviderHttpError(provider_rate_limited)', async () => {
    const http = new FakeHttpExecutor();
    http.queueResponse(429, '{"error":{"message":"rate limit"}}');
    const provider = new AnthropicProvider(config(), http);

    let captured: unknown = null;
    try {
      await provider.generatePlan({});
    } catch (err) {
      captured = err;
    }
    expect(captured).toBeInstanceOf(ProviderHttpError);
    expect((captured as ProviderHttpError).code).toBe('provider_rate_limited');
  });

  it('403 surfaces the upstream message intact (executor handles redaction)', async () => {
    const http = new FakeHttpExecutor();
    const key = 'sk-ant-test-ABCDEFGHIJKLMNOP';
    http.queueResponse(
      403,
      `{"type":"error","error":{"type":"forbidden","message":"key ${key} not allowed"}}`,
    );
    const provider = new AnthropicProvider(config(), http);

    let captured: unknown = null;
    try {
      await provider.generatePlan({});
    } catch (err) {
      captured = err;
    }
    expect(captured).toBeInstanceOf(ProviderHttpError);
    const err = captured as ProviderHttpError;
    expect(err.code).toBe('provider_auth');
    expect(err.message.includes('not allowed')).toBe(true);
  });
});

describe('AnthropicProvider.reviewPlan', () => {
  it('uses REVIEW max_tokens and sends the REVIEW system prompt verbatim', async () => {
    const http = new FakeHttpExecutor();
    http.queueResponse(
      200,
      JSON.stringify({
        content: [
          {
            type: 'text',
            text: '{"decision":"approve","reason":"ok","summary":"go","evidence":{}}',
          },
        ],
      }),
    );
    const provider = new AnthropicProvider(config(), http);
    const result = await provider.reviewPlan({ plan: {} });

    expect(result.decision).toBe('approve');
    const body = JSON.parse(http.calls[0]!.body) as Record<string, unknown>;
    expect(body.max_tokens).toBe(1800);
    expect(body.system).toBe(DEVICE_AGENT_SYSTEM_PROMPTS.REVIEW);
  });

  it('runs the two-pass flow when review research is needed: research call with web_search, then structured review without it', async () => {
    const http = new FakeHttpExecutor();
    // First response: research pass (web_search tool emits a research summary + citations).
    http.queueResponse(
      200,
      JSON.stringify({
        content: [
          {
            type: 'text',
            text: 'Helium Mobile Air Plan costs $15/month per the official Helium Mobile site.',
            citations: [{ url: 'https://heliummobile.com/plans', title: 'Helium Mobile - Plans' }],
          },
        ],
      }),
    );
    // Second response: structured review consuming the research evidence.
    http.queueResponse(
      200,
      JSON.stringify({
        content: [
          {
            type: 'text',
            text: '{"decision":"approve","reason":"$15/month is under $20.","summary":"ok","evidence":{"findings":[{"label":"Monthly rate","value":"$15/month","tone":"good"}]}}',
          },
        ],
      }),
    );
    const provider = new AnthropicProvider(config(), http);
    const payload = {
      plan: {},
      research: { needed: true, mode: 'auto_current_facts', currentDate: '2026-05-16T03:00:00.000Z', maxSearches: 2 },
    };
    const result = await provider.reviewPlan(payload);

    expect(result.decision).toBe('approve');
    expect(result.evidence).toMatchObject({
      research: { status: 'checked' },
      findings: expect.arrayContaining([
        { label: 'Current research', value: expect.stringContaining('Helium Mobile Air Plan'), tone: 'neutral' },
      ]),
      sources: expect.arrayContaining([
        { url: 'https://heliummobile.com/plans', title: 'Helium Mobile - Plans' },
      ]),
    });
    expect(http.calls).toHaveLength(2);

    // First call: research pass — web_search tool bound.
    const researchBody = JSON.parse(http.calls[0]!.body) as Record<string, unknown>;
    const researchTools = researchBody.tools as Array<Record<string, unknown>>;
    expect(researchTools[0]).toMatchObject({ type: 'web_search_20250305', name: 'web_search', max_uses: 2 });

    // Second call: structured review — NO web_search bound (research already done).
    const reviewBody = JSON.parse(http.calls[1]!.body) as Record<string, unknown>;
    expect(reviewBody.tools).toBeUndefined();
    const reviewMessages = reviewBody.messages as Array<{ role: string; content: string }>;
    const reviewUserContent = JSON.parse(reviewMessages[0]!.content) as Record<string, unknown>;
    // Research evidence has been embedded into context for the review pass.
    expect((reviewUserContent.context as Record<string, unknown>).researchEvidence).toBeDefined();
    expect((reviewUserContent.research as Record<string, unknown>).needed).toBe(false);
    expect((reviewUserContent.research as Record<string, unknown>).mode).toBe('provided_current_facts');
  });
});

describe('AnthropicProvider.ask', () => {
  it('uses ASK max_tokens, ASK temperature, and wraps response as { output_text }', async () => {
    const http = new FakeHttpExecutor();
    http.queueResponse(
      200,
      JSON.stringify({ content: [{ type: 'text', text: 'It swaps SOL for USDC.' }] }),
    );
    const provider = new AnthropicProvider(config(), http);
    const result = await provider.ask({ question: 'what happens?' });

    expect(result.output_text).toBe('It swaps SOL for USDC.');
    const body = JSON.parse(http.calls[0]!.body) as Record<string, unknown>;
    expect(body.max_tokens).toBe(800);
    expect(body.temperature).toBe(0.3);
    expect(body.system).toBe(DEVICE_AGENT_SYSTEM_PROMPTS.ASK);
  });
});

describe('AnthropicProvider via OpenRouter (browser CORS)', () => {
  function openRouterConfig(model = 'anthropic/claude-sonnet-4.5'): RuntimeConfig {
    return {
      provider: 'openrouter',
      apiFormat: 'openai-compatible',
      model,
      baseUrl: 'https://openrouter.ai/api/v1',
      apiKey: 'sk-or-test-ABCDEFGHIJKLMNOP',
    };
  }

  it('sends Bearer + OpenRouter attribution headers but OMITS X-OpenRouter-Metadata for the browser CORS preflight', async () => {
    const http = new FakeHttpExecutor();
    http.queueResponse(200, JSON.stringify({ content: [{ type: 'text', text: '{"intent":"x"}' }] }));
    const provider = new AnthropicProvider(openRouterConfig(), http);
    await provider.generatePlan({ userPrompt: 'swap 1 SOL' });

    const call = http.calls[0]!;
    expect(call.url).toBe('https://openrouter.ai/api/v1/messages');
    expect(call.headers.Authorization).toBe('Bearer sk-or-test-ABCDEFGHIJKLMNOP');
    // Documented OpenRouter browser headers ARE sent...
    expect(call.headers['HTTP-Referer']).toBeTruthy();
    expect(call.headers['X-Title']).toBeTruthy();
    // ...but the undocumented metadata header is omitted so it can't break the CORS preflight.
    expect('X-OpenRouter-Metadata' in call.headers).toBe(false);
    // Anthropic-direct headers are NOT sent on the OpenRouter route.
    expect('x-api-key' in call.headers).toBe(false);
    expect('anthropic-dangerous-direct-browser-access' in call.headers).toBe(false);
  });

  it('research pass binds Anthropic NATIVE web_search (NOT openrouter:web_search) on the /messages skin', async () => {
    const http = new FakeHttpExecutor();
    // Research pass: official citation present → grounded $15 answer flows through.
    http.queueResponse(200, JSON.stringify({
      content: [{
        type: 'text',
        text: 'Helium Mobile Air Plan is $15/month per the official Helium Mobile site.',
        citations: [{ url: 'https://heliummobile.com/plans', title: 'Helium Mobile - Plans' }],
      }],
    }));
    // Structured review pass.
    http.queueResponse(200, JSON.stringify({
      content: [{ type: 'text', text: '{"decision":"approve","reason":"$15 < $20","summary":"ok","evidence":{}}' }],
    }));
    const provider = new AnthropicProvider(openRouterConfig(), http);
    const result = await provider.reviewPlan({
      instruction: 'Approve only if the lowest helium monthly phone plan is less than $20',
      plan: {},
      research: { needed: true, mode: 'auto_current_facts', currentDate: '2026-06-09T00:00:00.000Z', maxSearches: 3 },
    });

    expect(result.decision).toBe('approve');
    expect(http.calls).toHaveLength(2);
    const researchCall = http.calls[0]!;
    expect(researchCall.url).toBe('https://openrouter.ai/api/v1/messages');
    const researchBody = JSON.parse(researchCall.body) as Record<string, unknown>;
    const tools = researchBody.tools as Array<Record<string, unknown>>;
    // The fix: native Anthropic server tool, which OpenRouter's Anthropic skin forwards to
    // Anthropic 1P. The old `openrouter:web_search` was silently dropped on /messages, leaving
    // Claude to answer ungrounded from training (the Helium "$0" bug).
    expect(tools[0]).toMatchObject({ type: 'web_search_20250305', name: 'web_search' });
    expect(researchCall.body).not.toContain('openrouter:web_search');
  });

  it('pricing research with NO citations falls back to "could not be verified" (never a fabricated price)', async () => {
    const http = new FakeHttpExecutor();
    // Research pass: model produced a price from training but returned ZERO citations — exactly
    // what happens when a web-search tool silently never runs (the OpenRouter+Claude "$0").
    http.queueResponse(200, JSON.stringify({
      content: [{ type: 'text', text: "Helium's Zero Plan is $0/month." }],
    }));
    // Structured review pass.
    http.queueResponse(200, JSON.stringify({
      content: [{ type: 'text', text: '{"decision":"needs_input","reason":"confirm price","summary":"verify","evidence":{}}' }],
    }));
    const provider = new AnthropicProvider(openRouterConfig(), http);
    await provider.reviewPlan({
      instruction: 'Approve only if the lowest helium monthly phone plan is less than $20',
      plan: {},
      research: { needed: true, mode: 'auto_current_facts', currentDate: '2026-06-09T00:00:00.000Z', maxSearches: 3 },
    });

    expect(http.calls).toHaveLength(2);
    const reviewBody = JSON.parse(http.calls[1]!.body) as Record<string, unknown>;
    const reviewMessages = reviewBody.messages as Array<{ role: string; content: string }>;
    const reviewUserContent = JSON.parse(reviewMessages[0]!.content) as Record<string, unknown>;
    const evidence = (reviewUserContent.context as Record<string, unknown>).researchEvidence as Record<string, unknown>;
    expect(evidence).toBeDefined();
    expect(String(evidence.summary)).toContain('could not be verified');
    // The ungrounded "$0" never reaches the review pass as a researched fact.
    expect(String(evidence.summary)).not.toContain('$0');
    expect(evidence.sources).toEqual([]);
  });
});

File: apps/android-twa/app/src/test/java/com/agentic/wallet/agent/provider/AnthropicProviderTest.kt
package com.agentic.wallet.agent.provider

import com.agentic.wallet.agent.prompts.DeviceAgentMessageAssembler
import com.agentic.wallet.agent.prompts.DeviceAgentSystemPrompts
import com.agentic.wallet.agent.runtime.RuntimeConfig
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.runBlocking
import org.json.JSONArray
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.time.Clock
import java.time.Instant
import java.time.ZoneOffset

class AnthropicProviderTest {
    private fun config(model: String = "claude-opus-4-5"): RuntimeConfig = RuntimeConfig(
        provider = "anthropic",
        apiFormat = "anthropic",
        model = model,
        baseUrl = "https://api.anthropic.com",
        apiKey = "sk-ant-test-ABCDEFGHIJKLMNOP",
        walletAddress = null,
    )

    @Test
    fun generatePlanSendsSystemAndUserRoles() = runBlocking {
        val http = FakeHttpExecutor().apply {
            queueResponse(
                200,
                """{"content":[{"type":"text","text":"{\"intent\":\"swap\",\"route\":\"jupiter\",\"risk\":\"low\",\"approval\":\"once\",\"safeguards\":[\"check slippage\"]}"}]}""",
            )
        }
        val provider = AnthropicProvider(config(), http)

        val result = provider.generatePlan(JSONObject().put("userPrompt", "swap 1 SOL"))

        assertEquals("swap", result.optString("intent"))
        val call = http.calls.single()
        assertEquals("https://api.anthropic.com/v1/messages", call.url)
        assertEquals("sk-ant-test-ABCDEFGHIJKLMNOP", call.headers["x-api-key"])
        assertEquals("2023-06-01", call.headers["anthropic-version"])
        // Phase 3 must not advertise itself as browser-direct
        assertFalse(call.headers.containsKey("anthropic-dangerous-direct-browser-access"))

        val body = JSONObject(call.body)
        assertEquals("claude-opus-4-5", body.optString("model"))
        assertEquals(1024, body.optInt("max_tokens"))
        assertEquals(0.2, body.optDouble("temperature"), 0.0001)
        assertTrue("system field must be top-level", body.has("system"))
        val messages = body.optJSONArray("messages")!!
        assertEquals(1, messages.length())
        assertEquals("user", messages.getJSONObject(0).getString("role"))
        // No web search tools
        assertFalse(body.has("tools"))
    }

    @Test
    fun reviewPlanUses1024MaxTokens() = runBlocking {
        val http = FakeHttpExecutor().apply {
            queueResponse(
                200,
                """{"content":[{"type":"text","text":"{\"decision\":\"approve\",\"reason\":\"ok\",\"summary\":\"go\",\"evidence\":{}}"}]}""",
            )
        }
        val provider = AnthropicProvider(config(), http)

        val result = provider.reviewPlan(JSONObject().put("plan", JSONObject()))

        assertEquals("approve", result.optString("decision"))
        val body = JSONObject(http.calls.single().body)
        assertEquals(1024, body.optInt("max_tokens"))
    }

    @Test
    fun reviewPlanRunsTwoPassResearchWhenResearchNeeded() = runBlocking {
        val http = FakeHttpExecutor().apply {
            // Pass 1: research call returns grounded text + citations.
            queueResponse(
                200,
                """{"content":[{"type":"text","text":"Helium Mobile cheapest plan is Air at ${'$'}15/month.","citations":[{"type":"web_search_result_location","url":"https://www.heliummobile.com/plans","title":"Plans"}]}]}""",
            )
            // Pass 2: structured review approves.
            queueResponse(
                200,
                """{"content":[{"type":"text","text":"{\"decision\":\"approve\",\"reason\":\"${'$'}15/month is under ${'$'}20.\",\"summary\":\"ok\",\"evidence\":{}}"}]}""",
            )
        }
        val fixedClock = Clock.fixed(Instant.parse("2026-05-18T00:00:00Z"), ZoneOffset.UTC)
        val provider = AnthropicProvider(config(), http, fixedClock)
        val payload = JSONObject()
            .put("instruction", "check helium mobile. lowest monthly plan. if less than ${'$'}20. approve.")
            .put("plan", JSONObject())
            .put(
                "research",
                JSONObject()
                    .put("needed", true)
                    .put("mode", "auto_current_facts")
                    .put("currentDate", "2026-05-16T03:00:00Z")
                    .put("maxSearches", 2),
            )

        val result = provider.reviewPlan(payload)

        assertEquals("approve", result.optString("decision"))
        assertEquals(2, http.calls.size)

        // Pass 1: web_search tool attached.
        val researchBody = JSONObject(http.calls[0].body)
        val tools = researchBody.optJSONArray("tools")!!
        assertEquals("web_search_20250305", tools.getJSONObject(0).optString("type"))
        assertEquals("web_search", tools.getJSONObject(0).optString("name"))
        assertEquals(2, tools.getJSONObject(0).optInt("max_uses"))

        // Pass 2: no tools; review payload carries the injected researchEvidence in the user content.
        val reviewBody = JSONObject(http.calls[1].body)
        assertFalse("review pass must not re-attach web_search", reviewBody.has("tools"))
        val userContent = reviewBody.optJSONArray("messages")!!.getJSONObject(0).optString("content")
        assertTrue("review user content must include researchEvidence", userContent.contains("researchEvidence"))
        assertTrue(userContent.contains("heliummobile.com"))
        assertTrue(
            "review user content must contain clock-fixed checkedAt; got: $userContent",
            userContent.contains("\"checkedAt\":\"2026-05-18T00:00:00Z\""),
        )
    }

    @Test
    fun reviewPlanFallsBackToSinglePassWhenResearchFails() = runBlocking {
        val http = FakeHttpExecutor().apply {
            queueFailure(RuntimeException("research pass network blip"))
            queueResponse(
                200,
                """{"content":[{"type":"text","text":"{\"decision\":\"needs_input\",\"reason\":\"Cannot confirm.\",\"summary\":\"ni\",\"evidence\":{}}"}]}""",
            )
        }
        val provider = AnthropicProvider(config(), http)
        val payload = JSONObject()
            .put("plan", JSONObject())
            .put(
                "research",
                JSONObject().put("needed", true).put("mode", "auto_current_facts"),
            )
        val result = provider.reviewPlan(payload)

        assertEquals("needs_input", result.optString("decision"))
        assertEquals(2, http.calls.size)
    }

    @Test
    fun reviewPlanFiltersBlogCitationsOnPricingQuestion() = runBlocking {
        val http = FakeHttpExecutor().apply {
            queueResponse(
                200,
                """{"content":[{"type":"text","text":"Helium Mobile cheapest plan is Air at ${'$'}15/month.","citations":[{"type":"web_search_result_location","url":"https://blog.heliummobile.com/break-free","title":"Break Free"},{"type":"web_search_result_location","url":"https://www.heliummobile.com/plans","title":"Plans"}]}]}""",
            )
            queueResponse(
                200,
                """{"content":[{"type":"text","text":"{\"decision\":\"approve\",\"reason\":\"ok\",\"summary\":\"ok\",\"evidence\":{}}"}]}""",
            )
        }
        val provider = AnthropicProvider(config(), http)
        provider.reviewPlan(
            JSONObject()
                .put("instruction", "check helium mobile. lowest monthly plan. if less than ${'$'}20. approve.")
                .put("plan", JSONObject())
                .put("research", JSONObject().put("needed", true).put("mode", "auto_current_facts")),
        )

        val reviewBody = JSONObject(http.calls[1].body)
        val userContent = reviewBody.optJSONArray("messages")!!.getJSONObject(0).optString("content")
        assertTrue(userContent.contains("heliummobile.com/plans"))
        assertFalse(userContent.contains("blog.heliummobile.com"))
    }

    @Test
    fun reviewPlanSuppressesSummaryWhenOnlyBlogCitationsForPricing() = runBlocking {
        val http = FakeHttpExecutor().apply {
            queueResponse(
                200,
                """{"content":[{"type":"text","text":"Helium Mobile offers the Zero Plan at ${'$'}0/month.","citations":[{"type":"web_search_result_location","url":"https://blog.heliummobile.com/zero-plan","title":"Zero Plan"}]}]}""",
            )
            queueResponse(
                200,
                """{"content":[{"type":"text","text":"{\"decision\":\"needs_input\",\"reason\":\"Not verified.\",\"summary\":\"ni\",\"evidence\":{}}"}]}""",
            )
        }
        val provider = AnthropicProvider(config(), http)
        provider.reviewPlan(
            JSONObject()
                .put("instruction", "check helium mobile. lowest monthly plan. if less than ${'$'}20. approve.")
                .put("plan", JSONObject())
                .put("research", JSONObject().put("needed", true).put("mode", "auto_current_facts")),
        )

        val reviewBody = JSONObject(http.calls[1].body)
        val userContent = reviewBody.optJSONArray("messages")!!.getJSONObject(0).optString("content")
        assertTrue(userContent.contains("Current pricing could not be verified"))
        assertFalse(userContent.contains("Zero Plan"))
        assertFalse(userContent.contains("blog.heliummobile.com"))
    }

    @Test
    fun cancellationDuringResearchPassPropagates() {
        val http = FakeHttpExecutor().apply {
            queueFailure(CancellationException("user backed out"))
        }
        val provider = AnthropicProvider(config(), http)
        val ex = runCatching {
            runBlocking {
                provider.reviewPlan(
                    JSONObject()
                        .put("plan", JSONObject().put("intent", "swap"))
                        .put(
                            "research",
                            JSONObject().put("needed", true).put("mode", "auto_current_facts"),
                        ),
                )
            }
        }.exceptionOrNull()
        assertTrue("Expected CancellationException, got $ex", ex is CancellationException)
        assertEquals(1, http.calls.size)
    }

    @Test
    fun singlePassReviewKeepsExistingShape() = runBlocking {
        val http = FakeHttpExecutor().apply {
            queueResponse(
                200,
                """{"content":[{"type":"text","text":"{\"decision\":\"approve\",\"reason\":\"ok\",\"summary\":\"ok\",\"evidence\":{}}"}]}""",
            )
        }
        val provider = AnthropicProvider(config(), http)
        provider.reviewPlan(
            JSONObject()
                .put("plan", JSONObject())
                .put(
                    "research",
                    JSONObject().put("needed", false).put("mode", "not_required"),
                ),
        )

        assertEquals(1, http.calls.size)
        val body = JSONObject(http.calls.single().body)
        assertFalse(body.has("tools"))
    }

    @Test
    fun askUses800MaxTokensAndWrapsResponseAsOutputText() = runBlocking {
        val http = FakeHttpExecutor().apply {
            queueResponse(
                200,
                """{"content":[{"type":"text","text":"It swaps SOL for USDC."}]}""",
            )
        }
        val provider = AnthropicProvider(config(), http)

        val result = provider.ask(JSONObject().put("question", "what happens?"))

        assertEquals("It swaps SOL for USDC.", result.optString("output_text"))
        val body = JSONObject(http.calls.single().body)
        assertEquals(800, body.optInt("max_tokens"))
        assertEquals(0.3, body.optDouble("temperature"), 0.0001)
    }

    @Test
    fun rateLimitMapsToRateLimitedCode() {
        val http = FakeHttpExecutor().apply {
            queueResponse(429, """{"error":{"message":"rate limit"}}""")
        }
        val provider = AnthropicProvider(config(), http)
        val ex = try {
            runBlocking { provider.generatePlan(JSONObject()) }
            null
        } catch (e: ProviderHttpException) {
            e
        }
        assertEquals(ProviderErrorCodes.RATE_LIMITED, ex!!.code)
    }

    @Test
    fun systemFieldOnWireIsVerbatimPlanPrompt() = runBlocking {
        val http = FakeHttpExecutor().apply {
            queueResponse(200, """{"content":[{"type":"text","text":"{\"intent\":\"x\"}"}]}""")
        }
        val provider = AnthropicProvider(config(), http)
        provider.generatePlan(JSONObject().put("userPrompt", "do thing"))

        val body = JSONObject(http.calls.single().body)
        assertEquals(DeviceAgentSystemPrompts.PLAN, body.getString("system"))
    }

    @Test
    fun systemFieldOnWireIsVerbatimReviewPrompt() = runBlocking {
        val http = FakeHttpExecutor().apply {
            queueResponse(200, """{"content":[{"type":"text","text":"{\"decision\":\"approve\",\"reason\":\"ok\",\"summary\":\"go\",\"evidence\":{}}"}]}""")
        }
        val provider = AnthropicProvider(config(), http)
        provider.reviewPlan(JSONObject().put("plan", JSONObject()))

        val body = JSONObject(http.calls.single().body)
        assertEquals(DeviceAgentSystemPrompts.REVIEW, body.getString("system"))
    }

    @Test
    fun systemFieldOnWireIsVerbatimAskPrompt() = runBlocking {
        val http = FakeHttpExecutor().apply {
            queueResponse(200, """{"content":[{"type":"text","text":"the answer"}]}""")
        }
        val provider = AnthropicProvider(config(), http)
        provider.ask(JSONObject().put("question", "?"))

        val body = JSONObject(http.calls.single().body)
        assertEquals(DeviceAgentSystemPrompts.ASK, body.getString("system"))
    }

    @Test
    fun userContentMatchesAssemblerOutput() = runBlocking {
        val http = FakeHttpExecutor().apply {
            queueResponse(200, """{"content":[{"type":"text","text":"{\"intent\":\"x\"}"}]}""")
        }
        val provider = AnthropicProvider(config(), http)
        val payload = JSONObject()
            .put("userPrompt", "swap 1 SOL")
            .put("userNotes", "make it fast")
        provider.generatePlan(payload)

        val expected = DeviceAgentMessageAssembler.buildPlanMessages(payload).userContent
        val body = JSONObject(http.calls.single().body)
        val actual = body.getJSONArray("messages").getJSONObject(0).getString("content")
        assertEquals(expected, actual)
    }

    @Test
    fun httpErrorBodyEchoingKeyIsRedactedAtProviderLayer() {
        val key = "sk-ant-test-ABCDEFGHIJKLMNOP"
        val http = FakeHttpExecutor().apply {
            queueResponse(403, """{"type":"error","error":{"type":"forbidden","message":"key $key not allowed"}}""")
        }
        val provider = AnthropicProvider(config(), http)
        val ex = try {
            runBlocking { provider.generatePlan(JSONObject()) }
            null
        } catch (e: ProviderHttpException) {
            e
        }
        assertNotEquals(null, ex)
        assertEquals(ProviderErrorCodes.AUTH, ex!!.code)
        assertTrue("Provider should surface the upstream message", ex.message.contains("not allowed"))
    }
}
